File: src/base/linalg.cpp
#include "linalg.h"

namespace Linalg {

/** 
 * @brief Perform dot product: ret = transpose(x) * y
 *
 * @param size      Vector length
 * @param x         First Vector ptr
 * @param y         Second Vector ptr
 * @return f64   transpose(vec1) * vec2
 */
f64 dot(const int size, const f64* x, const f64* y) {
    f64 ret = 0.0;
    for (int i = 0; i < size; i++) {
        ret += x[i] * y[i];
    }
    return ret;
}

/** 
 * @brief Perform inplace squaring (x_1^2, x_2^2, ..., x_size^2)
 *
 * @param size      Vector length
 * @param x         Vector ptr
 */
void square(const int size, f64* x) {
    for (int i = 0; i < size; i++) {
        x[i] *= x[i];
    }
}


void diagmat_vec(const f64* D, bool invD, const f64* x, const int size, f64* out) {
    int i;

    if (invD) {
        for (i = 0; i < size; i++) {
            out[i] = x[i] / D[i];
        }
    }
    else {
        for (i = 0; i < size; i++) {
            out[i] = D[i] * x[i];
        }
    }
}

void diagmat_vec_inplace(const f64* D, bool invD, f64* x, const int size) {
    int i;

    if (invD) {
        for (i = 0; i < size; i++) {
            x[i] /= D[i];
        }
    }
    else {
        for (i = 0; i < size; i++) {
            x[i] *= D[i];
        }
    }
}

/** 
 * @brief Perform diagonal scaled axpy: out = D * (x + beta * y) or out = D^(-1) * (x + beta * y)
 *
 * @param size      Vector length
 * @param x         Vector x
 * @param y         Vector y
 * @param D         Diagonal Matrix D
 * @param beta      Scaling factor of beta
 * @param invD      Invert Matrix D
 * @param out       Vector to fill
 */
void dsaxpy(const int size, const f64* x, const f64* y, const f64* D, f64 beta, bool invD, f64* out) {
    int i;

    if (invD) {
        for (i = 0; i < size; i++) {
            out[i] = (x[i] + beta * y[i]) / D[i];
        }
    }
    else {
        for (i = 0; i < size; i++) {
            out[i] = D[i] * (x[i] + beta * y[i]);
        }
    }
}

/** 
 * @brief Perform diagonal general matrix vector: out = D * x + beta * y or D^(-1) * x + beta * y
 *
 * @param size      Vector length
 * @param x         Vector x
 * @param y         Vector y
 * @param D         Diagonal Matrix D
 * @param beta      Scaling factor of beta
 * @param invD      Invert Matrix D
 * @param out       Vector to fill
 */
void dgmv(const int size, const f64* x, const f64* y, const f64* D, f64 beta, bool invD, f64* out) {
    int i;

    if (invD) {
        for (i = 0; i < size; i++) {
            out[i] = x[i] / D[i] + beta * y[i];
        }
    }
    else {
        for (i = 0; i < size; i++) {
            out[i] =  D[i] * x[i] + beta * y[i];
        }
    }
}

} // namespace Linalg

File: bindings/python/test/linalg.py
import moopy.linalg as linalg

def dot():
    return linalg.dot([1, 2], [3, 4]) == 11

def square():
    x = [1.0, 2.0]
    x = linalg.square(x)
    return (x[0] == 1.0) and (x[1] == 4.0)

def matrix_vector():
    A = [1.0, 2.0, 3.0, 4.0]
    x = [5.0, 6.0]
    b = linalg.matrix_vector('R', A, x)
    return (b[0] == 17.0) and (b[1] == 39.0)    

def diagmat_vec():
    D = [1.0, 2.0]
    x = [4.0, 5.0]
    b = linalg.diagmat_vec(D, False, x)
    return (b[0] == 4.0) and (b[1] == 10.0)

def dsaxpy():
    x = [1.0, 2.0]
    beta = 2.0
    y = [4.0, 5.0]
    D = [3.0, -3.0]
    y_out = linalg.dsaxpy(x, y, D, beta, False)
    return (y_out[0] == 27.0) and (y_out[1] == -36.0)

def dgmv():
    x = [5.0, 6.0]
    y = [5.0, 6.0]
    D = [7.0, 8.0]
    beta = -2.0
    b = linalg.dgmv(x, y, D, beta, False)
    return (b[0] == 25.0) and (b[1] == 36.0)

print(dot())
print(square())
print(matrix_vector())
print(diagmat_vec())
print(dsaxpy())
print(dgmv())
